### ISSUE TITLE
HAL serializer doesn't respect skip_null_values=false for relations

### ISSUE THREAD

**Issue by carlobeltrame** [api-platform/core#4372]
**API Platform version(s) affected**: 2.6.5, and any other API platform version using Symfony 4.2 (since the [introduction](https://github.com/symfony/symfony/pull/28661) of the `skip_null_values` flag)

**Description**  
The HAL format normalizer does things a little differently from the rest of the normalizers, especially when it comes to normalizing relations. In [ItemNormalizer#populateRelation](https://github.com/api-platform/core/blob/main/src/Hal/Serializer/ItemNormalizer.php#L192), an `empty()` check determines whether to leave out a relation. This also affects optional ToOne relations which can have a null value. In the rest of the framework, the `skip_null_values` flag in the context determines what to do with such null relations. In HAL, this flag is not respected.

I'd be glad to provide a PR to fix this if requested.

**How to reproduce**  
Create two entities like the following:
```php
/**
 * @ORM\Entity
 */
#[ApiResource(normalizationContext: ['skip_null_values' => false])]
class Dummy {
  public ?RelatedDummy $relatedDummy = null;
}

/**
 * @ORM\Entity
 */
#[ApiResource]
class RelatedDummy {
}
```
Activate support for HAL in the config:
```yml
api_platform:
    formats:
        jsonhal: [ 'application/hal+json' ]
        jsonld: [ 'application/ld+json' ]
```
Create a `Dummy` instance in the database without any `RelatedDummy`, and fetch it using the format `Accept: application/hal+json`.
Expected result:
```json
{
  "relatedDummy": null
}
```
Actual result:
```json
{}
```

**Possible Solution**  
Use a check like the following:
```php
if (empty($attributeValue)) {
    $skipNullValues = $context[self::SKIP_NULL_VALUES] ?? $this->defaultContext[self::SKIP_NULL_VALUES] ?? false;
    if (null !== $attributeValue || $skipNullValues) {
        continue;
    }
}
```

**Comment by dunglas** [comment]
The fix you propose looks good to me. Would you like to work on a patch?

**Comment by carlobeltrame** [comment]
I have created a PR #4373 for demonstration, but I have run into problems. The [hyperschema](https://hyperschema.org/mediatypes/hal/) does not allow `null` entries in `_links` and `_embedded`. I have seen many places on the internet where people have been confused by what is and isn't allowed in HAL concerning null values:
* An unresolved conversation from 2011 on [hal-discuss](https://groups.google.com/g/hal-discuss/c/tdVTMux9pwI) where the HAL spec author didn't reply anymore, but it kind of sounds like it should be allowed (?)
* Zend Expressive / Mezzio had an [issue](https://github.com/zendframework/zend-expressive-hal/issues/52) on this topic, and they serialize the `null` value outside of the `_links` and `_embedded`, next to the primitive properties
* Spring Data Rest circumvents the problem by specifying any relations as [subresource links](https://github.com/spring-projects/spring-data-rest/issues/1774), which are always rendered but may return a 404
* Nevertheless, the SDR maintainer Oliver Drotbohm makes an important [argument](https://stackoverflow.com/a/24514621) from a HATEOAS perspective: If the null links are simply omitted, how would a client know of the possibility to assign something to that relation? In the case of API platform, we have the OpenAPI spec, but in a broader sense, the HAL specification kind of needs to allow null links for that reason, doesn't it?
* Some other [unknown API implementations](https://github.com/carlobeltrame/hal-json-normalizer/pull/6) also use `null` in links and embedded, but are "aware" of the "inconsistencies"
* The HAL [spec](https://datatracker.ietf.org/doc/html/draft-kelly-json-hal-08) itself doesn't mention `null` a single time, and mentions that entries in `_links` should be "either a Link Object or an array of Link Objects", and entries in `_embedded` should be "either a Resource Object or an array of Resource Objects".

So, to proceed, I see multiple options.
1. We could stay with the old implementation and just ignore `skip_null_values: false`, effectively pushing the null checking responsibility to any API clients:
```json
{
  "_links": {
    "self": { "href": "/some/iri" }
  }
}
```
2. We could deviate from the upstream hyperschema and use the format I implemented in #4373, which would follow the rest of API platform's supported formats the closest:
```json
{
  "_links": {
    "self": { "href": "/some/iri" },
    "relatedDummy": null
  },
  "_embedded": {
    "relatedDummy": null
  }
}
```
3. We could choose to only null the `href` of the null links, which is technically also incorrect but may be a little less problematic for clients:
```json
{
  "_links": {
    "self": { "href": "/some/iri" },
    "relatedDummy": { "href": null }
  },
  "_embedded": {
    "relatedDummy": null
  }
}
```
4. We could render the `null` value outside of `_links` and `_embedded`, like Mezzio does, assuming that clients will merge the three categories together anyways:
```json
{
  "_links": {
    "self": { "href": "/some/iri" }
  },
  "relatedDummy": null
}
```

To summarize:
| Option | 1 | 2 | 3 | 4
| -------- | --- | --- | --- | ---
| Compatible with spec | :heavy_check_mark: | :question: | :question: | :heavy_check_mark: 
| Compatible with hyperschema | :heavy_check_mark: | :x: | :x: | :heavy_check_mark: 
| Respects `skip_null_values` | :x: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: 
| Effort in API client to consume; amount of out-of-band information needed | high | low | low | medium 

What do you think? Which way would you choose?

**Comment by BacLuc** [comment]
I made another attempt for this issue with:
- https://github.com/api-platform/core/pull/7317

It does not fix the issue, but allows the user to enable it.

We have a composer patch in place since years:
- https://github.com/ecamp/ecamp3/pull/2017/files

And this can break of course:
- https://github.com/ecamp/ecamp3/commit/3032bd48e0525527a265bf172836ed054fdef1ea#diff-fa136484b0ea62d32b42eee33e34d9c53523b7444dff3d309c47182a440af5a0

So i would like to contribute our patch to upstream.
